core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ManifestEvaluator;
+import org.apache.iceberg.expressions.Projections;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.iceberg.util.TableScanUtil;
+
+abstract class DistributedDataBatchScan
+    extends SnapshotScan<BatchScan, ScanTask, ScanTaskGroup<ScanTask>> implements BatchScan {
+
+  private static final int NUM_LOCAL_CORES = Runtime.getRuntime().availableProcessors();
+  private static final long METADATA_LIMIT_PER_LOCAL_CORE = 128 * 1024 * 1024; // 128 MB
+  private static final int NUM_WORKER_THREADS = 2;
+
+  protected DistributedDataBatchScan(Table table, Schema schema, TableScanContext context) {
+    super(table, schema, context);
+  }
+
+  protected abstract int remoteParallelism();
+
+  protected abstract PlanningMode dataPlanningMode();
+
+  protected abstract List<DataFile> planDataRemotely(List<ManifestFile> manifests);
+
+  protected abstract PlanningMode deletePlanningMode();
+
+  protected abstract List<DeleteFile> planDeletesRemotely(List<ManifestFile> manifests);
+
+  protected abstract void releaseResources();
+
+  @Override
+  protected boolean useSnapshotSchema() {
+    return true;
+  }
+
+  @Override
+  protected CloseableIterable<ScanTask> doPlanFiles() {
+    Snapshot snapshot = snapshot();
+
+    List<ManifestFile> dataManifests = findMatchingDataManifests(snapshot);
+    List<ManifestFile> deleteManifests = findMatchingDeleteManifests(snapshot);
+
+    boolean planDataRemotely = shouldPlanRemotely(dataPlanningMode(), dataManifests);
+    boolean planDeletesRemotely = shouldPlanRemotely(deletePlanningMode(), deleteManifests);
+
+    if (planDataRemotely || planDeletesRemotely) {
+      ExecutorService workerPool = newWorkerPool();
+
+      CompletableFuture<CloseableIterable<DataFile>> dataFilesFuture =
+          newDataFilesFuture(dataManifests, planDataRemotely, workerPool);
+
+      CompletableFuture<DeleteFileIndex> deletesFuture =
+          newDeletesFuture(deleteManifests, planDeletesRemotely, workerPool);
+
+      try {
+        CloseableIterable<DataFile> dataFiles = dataFilesFuture.join();
+        DeleteFileIndex deletes = deletesFuture.join();
+        return createTasks(dataFiles, deletes);
```

Comment:
This has a performance issue. We look up the delete index in a single thread. I switched to `ParallelIterable` locally to assign deletes using multiple threads.